### PR TITLE
Updated lazy pagination

### PR DESCRIPTION
**Pull request by stephenwf**
@thewilkybarkid @giorgiosironi

### CODE REVIEW

File: src/CanBeSliced.php [hunk 1]
```diff
@@ -14,21 +14,17 @@
...
+        if (empty($this->pages[$page])) {
+            $this->pages[$page] = $this->slice($this->pageBatch * ($page - 1), $this->pageBatch);
         }
 
         if (false === isset($this->pages[$page])) {
             throw new LogicException('Could not find page '.$page);
         }
 
-        return iterator_to_array($this->pages[$page]);
+        return $this->pages[$page]->toArray();
```
Comment (stephenwf) [inline src/CanBeSliced.php]: @giorgiosironi @thewilkybarkid 

It looks to be consistently a sequence returned, the toArray just unwraps the underlying array. Would there be a knock on effect to just return a sequence and let the consumer convert to array if they need it? (Tried and tests still pass)

File: src/CanBeSliced.php [hunk 2]
```diff
@@ -14,21 +14,17 @@
 
     abstract public function slice(int $offset, int $length = null) : Sequence;
 
-    final private function getPage(int $page) : array
+    final private function getPage(int $page)
```
Comment (giorgiosironi) [inline src/CanBeSliced.php]: missing return type hint, shouldn't be necessary to remove it now?

Comment (stephenwf) [reply src/CanBeSliced.php]: Ah this is left over from my other comment about sequences, I will re-add based on that

File: src/CanBeSliced.php (hunk 1 above)
Comment (giorgiosironi) [reply src/CanBeSliced.php]: so slice() is returning a type-hinted `Sequence` which extends `Collection` which has `toArray()`. The only call is in `SlicedIterator`, and a `Sequence` is an `ArrayAccess` anyway; I'd say we can return a `Sequence`.

File: src/ArrayFromIterator.php [hunk 3]
```diff
@@ -43,11 +43,7 @@
     final private function all() : PromiseSequence
     {
         return new PromiseSequence(
-            $promise = new Promise(
-                function () use (&$promise) {
-                    $promise->resolve(new ArraySequence($this->toArray()));
-                }
-            )
+            new FulfilledPromise(new ArraySequence($this->toArray()))
```
Comment (giorgiosironi) [inline src/ArrayFromIterator.php]: I'm not sure what the original code was doing?

Comment (giorgiosironi) [reply src/ArrayFromIterator.php]: you may even be able to pass a promise of an array, the constructor of `PromiseSequence` seems to accept it

**Comment by giorgiosironi** [comment]
Is there a test we can add that fails with the old implementation and that passes now? Will document why we are doing this.

**Comment by stephenwf** [comment]
@giorgiosironi I've added a test that will fail on the old one

File: test/Client/ArticlesTest.php [hunk 4]
```diff
@@ -53,6 +53,24 @@ public function it_can_be_traversed()
     /**
      * @test
      */
+    public function it_does_not_make_unnecessary_network_requests()
```
Comment (giorgiosironi) [inline test/Client/ArticlesTest.php:56]: ..._to_pages_besides_the_ones_being_loaded

**Review by giorgiosironi — APPROVED** [review]

**Comment by giorgiosironi** [comment]
Also cannot reproduce.

**Comment by giorgiosironi** [comment]
Ok, now I can reproduce the original behavior which I wasn't seeing anymore after switching to a synchronous client, on `test3.php`.

**Comment by giorgiosironi** [comment]
In the tests we are using `MockMiddleware`, could be related.

File: src/ArrayFromIterator.php [hunk 5]
```diff
@@ -43,11 +42,7 @@
     final private function all() : PromiseSequence
     {
         return new PromiseSequence(
-            $promise = new Promise(
-                function () use (&$promise) {
-                    $promise->resolve(new ArraySequence($this->toArray()));
-                }
-            )
+            new FulfilledPromise($this->toArray())
```
Comment (thewilkybarkid) [inline src/ArrayFromIterator.php:45]: The original code stopped `toArray()` (and hence requests) from being immediately called.

File: src/CanBeSliced.php [hunk 6]
```diff
@@ -14,21 +14,17 @@
...
-            for ($i = 0; $i < $this->count(); ++$i) {
-                if (0 === $i % $this->pageBatch) {
-                    $this->pages[count($this->pages) + 1] = $this->slice($i, $this->pageBatch);
-                }
-            }
+        if (empty($this->pages[$page])) {
+            $this->pages[$page] = $this->slice($this->pageBatch * ($page - 1), $this->pageBatch);
         }
 
         if (false === isset($this->pages[$page])) {
```
Comment (thewilkybarkid) [inline src/CanBeSliced.php:23]: This wouldn't happen anymore.

**Comment by thewilkybarkid** [comment]
Don't think this is the right solution. What's happening before is that the pages are loaded asynchronously. Now, each page is requested sequentially. If you're only using the snippets (ie not then lazy-loading the remaining bits of the models), you're now _worse_ off (until we hit maybe a hundred pages). What we should look at doing it batching requests (so send 10 together, then the next 10, rather than all at once). This should happen in the API Client to, rather than the SDK.

File: src/CanBeSliced.php [hunk 7]
```diff
@@ -14,21 +14,17 @@
...
             throw new LogicException('Could not find page '.$page);
         }
 
-        return iterator_to_array($this->pages[$page]);
+        return $this->pages[$page];
```
Comment (thewilkybarkid) [inline src/CanBeSliced.php:27]: 👍 (was needed before `Sequence` implemented `ArrayAccess`)

**Comment by giorgiosironi** [comment]
What is the specification for the number of requests that should be performed when the SDK loads the first page (which should be rather common)? Does that mean that the default behavior when looking at page 1 is always to prefetch 10 pages, or to prefetch all of them?

**Comment by thewilkybarkid** [comment]
Loading all of them is happening because you're going to the whole set (and they should be batched). If you want a subset, use `slice()`.

**Comment by giorgiosironi** [comment]
Ok with `slice()` we should be safe from projects using the SDK DOSing other backends, while the full page loading behavior will hit them but only with a batch of requests at a time. The batching should also help with memory requirements, but we will test this out.

**Comment by giorgiosironi** [comment]
So I'd say not much to save from this attempt, we will reopen this to test correct behavior later.

**Comment by thewilkybarkid** [comment]
Going to close this.

(The change of `getPage()` return type to `Sequence` is good, but should be done separately.)

**Pull request closed without merging**